File: Creational Patterns/Prototype/Swift/Prototype.swift
//
//  Prototype.swift
//  
//
//  Created by Антон Тимонин on 15.02.2020.
//

import UIKit


// MARK:- Prototype
public protocol Shape {
    init(_ prototype: Self)
    
    var x: Int { get }
    var y: Int { get }
}

extension Shape {
    public func clone() -> Self {
        return type(of: self).init(self)
    }
}


class Rectangle: Shape {
    var x: Int
    var y: Int
    var color: String
    
    init(x: Int, y: Int, color: String) {
        self.x = x
        self.y = y
        self.color = color
    }
    
    required convenience init(_ prototype: Rectangle) {
        self.init(x: prototype.x, y: prototype.y, color: prototype.color)
    }
}


class Circle: Shape {
    var x: Int
    var y: Int
    var size: String

    init(x: Int, y: Int, size: String) {
        self.x = x
        self.y = y
        self.size = size
    }
    
    required convenience init(_ prototype: Circle) {
        self.init(x: prototype.x, y: prototype.y, size: prototype.size)
    }
}

File: Creational Patterns/Prototype/CS/Prototype.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Prototype
{
    interface IFigure
    {
        public IFigure clone();
        public double GetShapeArea();
        public void GetInfo();
    }
    
    class Circle : IFigure
    {
        private uint _radiusCircle;
        
        public Circle(uint radius)
        {
            this._radiusCircle = radius;
        }
        
        public IFigure clone()
        {
            return new Circle(this._radiusCircle);
        }

        public double GetShapeArea()
        {
            return 3.14 * this._radiusCircle * this._radiusCircle;
        }

        public void GetInfo()
        {
            Console.WriteLine("Shape of this Circle: " +  this._radiusCircle.ToString());
        }
    }

    class Rectangle: IFigure
    {
        private uint _w, _h;

        public Rectangle(uint w, uint h)
        {
            this._w = 2;
            this._h = h;
        }

        public IFigure clone()
        {
            return new Rectangle(this._w, this._h);
        }

        public double GetShapeArea()
        {
            return this._w * this._h;
        }
        
        public void GetInfo()
        {
            Console.WriteLine("Shape of this Rectangle: " + this._w + "x" + this._h);
        }
    }
}
